d# Homework 2 (Due: 2021.04.01 11:59 PM)

Name: 임정한

Student ID: 2017312172

For this homework, you are provided with an input file.
Submit the code that you used for analysis as an ipython notebook and an exported PDF file.

# Q1 Analyzing the Election Data (3 x 8 = 24 pts)

Run the following cell to set the filename.

In [92]:
filename = 'bush-gore-results-fl_demo.csv'

(a) Determine how many counties Bush won. You can assume that each row corresponds to a unique county.
    Demo file: 'bush-gore-results-fl_demo.csv' 

In [93]:
import pandas as pd
import numpy as np
df = pd.read_csv(filename)
#All possible candidates are 'bush', 'gore', 'brow', 'nade', 'harr', 'hage', 'buch', 'mcre', 'phil', 'moor'. You can find a person from the list.
df_resize = df.loc[:, ['bush', 'gore', 'brow', 'nade', 'harr', 'hage', 'buch', 'mcre', 'phil', 'moor']]

bush_won = 0
for length in range(df_resize.shape[0]):
    if df_resize.loc[length,:].max() == df_resize.loc[length,'bush']:
        bush_won += 1
print(bush_won)

51


(b) Determine who won the largest county. The column named "npop" records the size of each county.
    Demo file: 'bush-gore-results-fl_demo.csv' 

In [94]:
import pandas as pd
import numpy as np
#default
filename = 'bush-gore-results-fl_demo.csv'
df = pd.read_csv(filename)

max_index = df['npop'].argmax()


if (df.loc[max_index,'bush']<df.loc[max_index, 'gore']):
    print("Gore")
else:
    print("Bush")


Gore


(c) Determine the average number of votes per county that Buchanan obtained. You can assume that each row corresponds to a unique county.
    Demo file: 'bush-gore-results-fl_demo.csv' 

In [95]:
import pandas as pd
import numpy as np
#default
filename = 'bush-gore-results-fl_demo.csv'
df = pd.read_csv(filename)

df_resize = df.loc[:, 'buch']
#print(df_resize)
print(df_resize.mean())

360.67164179104475


(d) Determine, for the number of votes Buchanan obtained in Palm Beach, how many standard deviations it is away from Buchanan's overall mean, in absolute value. The row with coounty number 50 ("co"=50) records the results for Palm Beach County. (You can assumed that such a row will exist in the test case.)
    Demo file: 'bush-gore-results-fl_demo.csv' 

In [96]:
#It asks to compute (the absolute difference / std).

import pandas as pd
import numpy as np
#default
filename = 'bush-gore-results-fl_demo.csv'
df = pd.read_csv(filename)
co = 50

df_resize = df.loc[:, 'buch']
#print(df_resize)
buch_mean = df_resize.mean()
buch_std = df_resize.std()
print((df_resize.iloc[co-1]-buch_mean)/buch_std)
#print(np.square(df_resize.iloc[co-1]-buch_mean))



6.993018491215867


(e) Now calculate the above statistic (same as in part f) for all the counties and report them in decreasing order.
    Demo file: 'bush-gore-results-fl_demo.csv' 
    Example output: 
    county_50 6.993018  
    county_52 ...
    ...
    ...

In [97]:
import pandas as pd
import numpy as np
#default
filename = 'bush-gore-results-fl_demo.csv'
df = pd.read_csv(filename)


dict_var = {}
df_resize = df.loc[:, 'buch']

#print(df_resize)
buch_mean = df_resize.mean()
buch_std = df_resize.std()
for i in range(df.shape[0]):
    buch_value = df_resize.iloc[i]
    dict_var[i] = ((buch_value-buch_mean)/buch_std)
sorted_var = sorted(dict_var.items(), reverse = True, key = lambda x: x[1])
for index, values in enumerate(sorted_var):
    print("county_", values[0]+1, "\t", values[1])

county_ 50 	 6.993018491215867
county_ 52 	 1.6721223983806142
county_ 28 	 1.3031713743995565
county_ 6 	 1.1720381791291807
county_ 15 	 0.8697650510483141
county_ 5 	 0.6875121355877918
county_ 51 	 0.6875121355877918
county_ 41 	 0.6719539598777472
county_ 43 	 0.6652861702877281
county_ 53 	 0.6030534674475496
county_ 16 	 0.5363755715473585
county_ 64 	 0.5230399923673202
county_ 48 	 0.41191016586700174
county_ 55 	 0.11185963431614161
county_ 35 	 0.09852405513610338
county_ 56 	 0.09852405513610338
county_ 34 	 0.06296251065600145
county_ 36 	 0.04740433494595685
county_ 40 	 0.02295577311588676
county_ 9 	 0.02073317658588039
county_ 46 	 0.014065386995861275
county_ 1 	 0.00517500087583579
county_ 3 	 -0.028163947074259777
county_ 26 	 -0.04149952625429801
county_ 58 	 -0.07039328114438083
county_ 57 	 -0.14818415969460383
county_ 10 	 -0.16596493193465478
county_ 8 	 -0.17485531805468027
county_ 54 	 -0.2504236000748969
county_ 49 	 -0.257091389664916
county_ 27 	 -0.297098

(f) Assuming that the votes were distributed across the white, black, and hispanic population uniformly, determine which candidate obtained the largest number of votes for each subpopulation.
    Demo file: 'bush-gore-results-fl_demo.csv' 
    Example output:
    white: bush
    black: ..
    hispanic: ..

In [98]:
import pandas as pd
import numpy as np
#default
filename = 'bush-gore-results-fl_demo.csv'
df = pd.read_csv(filename)

df_feature = df.loc[:, ['whit','blac','hisp']]
candidate = ['bush', 'gore']
df_cand = df.loc[:, candidate]
cand_feat = [[0,0,0], [0,0,0]]

feature = ['whit', 'blac', 'hisp']
for i in range(df_cand.shape[0]):
    for j in range(3):
       cand_feat[0][j] += df_feature.iloc[i, j] / 100 * df_cand.iloc[i, 0]
       cand_feat[1][j] += df_feature.iloc[i, j] / 100 * df_cand.iloc[i, 1]

cand_feat = np.asarray(cand_feat)
for i in range(3):
    print(feature[i], " ", candidate[cand_feat.argmax(axis=0)[i]])

whit   bush
blac   gore
hisp   gore


(g) Calculate the correlation between the difference in votes between Bush and Gore, and the votes obtained by Nader.
(FYI: Pearson's correlation coefficient) https://en.wikipedia.org/wiki/Pearson_correlation_coefficient
    Demo file: 'bush-gore-results-fl_demo.csv' 

In [99]:
#(g) Calculate the correlation between the difference in votes between Bush and Gore, and the votes obtained by Nader.
#
import pandas as pd
import numpy as np
#default
filename = 'bush-gore-results-fl_demo.csv'
df = pd.read_csv(filename)
data = np.zeros(shape=(df.shape[0],2))
df_corr = pd.DataFrame(data, columns = ['diff', 'nade'])
df_corr['diff'] = df.loc[:, 'bush']-df.loc[:, 'gore']
df_corr['nade'] = df.loc[:, 'nade']

print(df_corr.corr().iloc[0,1])

-0.42364341310857706


(h) Find the distance between the county that Bush won by the largest margin and the county that Gore won by the largest margin. (Just use basic Euclidean distance between the latitude (lat) and longitude (lon) values for the counties, no need to compute spherical distance.) (FYI: Eucliden distane is decribed in https://en.wikipedia.org/wiki/Euclidean_distance#:~:text=In%20mathematics%2C%20the%20Euclidean%20distance,metric%20as%20the%20Pythagorean%20metric.)
    Demo file: 'bush-gore-results-fl_demo.csv' 

In [100]:
#(h) Find the distance between the county that Bush won by the largest margin and the county that Gore won by the largest margin. (Just use basic Euclidean distance between the latitude (lat) and longitude (lon) values for the counties, no need to compute spherical distance.) (FYI: Eucliden distane is decribed in https://en.wikipedia.org/wiki/Euclidean_distance#:~:text=In%20mathematics%2C%20the%20Euclidean%20distance,metric%20as%20the%20Pythagorean%20metric.)

import pandas as pd
import numpy as np
#default
filename = 'bush-gore-results-fl_demo.csv'
df = pd.read_csv(filename)
df_resize = df.loc[:, ['bush', 'gore', 'brow', 'nade', 'harr', 'hage', 'buch', 'mcre', 'phil', 'moor']]
df_resize['bush-2']=df_resize.loc[:, 'bush']-df_resize.apply(lambda row : row.nlargest(2).values[-1], axis=1)
df_resize['gore-2']=df_resize.loc[:, 'gore']-df_resize.apply(lambda row : row.nlargest(2).values[-1], axis=1)

bushidx = df_resize['bush-2'].idxmax()
goreidx = df_resize['gore-2'].idxmax()

latdiff = (np.square(df.loc[bushidx, 'lat']-df.loc[goreidx,'lat']))
londiff = (np.square(df.loc[bushidx, 'lon']-df.loc[goreidx,'lon']))
print(np.sqrt(latdiff+londiff))


4.341658669218476
